### PR TITLE
Settings: Add news options and fix some things

### PR DESCRIPTION
**Pull request by Zangetsu38**
- Fix save change on init. (On master before, save only if you set one think special, on last master it save also if you open settings dialog, causing problem for pref path)
- Add pref path in Emulator Tab for can set personal folder without edit config file manually.
- Fix log level choice works on config, on master it reset your value changed all time.
- Add log level choice in settings dialog.
- Add Background image path on setting dialog.
- Fix Background Alpha with disable it if not have image set.
- Add possibility can change background image without reboot.
- Add game background on font by @1whatleytay thx for that.
- Add gamebackfront apear with get mouse on icon.
- Add restore background user with set cursor on other Icon/title ID
(for it, i have added space free(TitleID), for not get game background or return background user by acident)
- Add Special path for Game background if it not found majority hb use it.
- Add option for can enable/disable viewing game background.

# Result  for pref path:
- By default value : 
image
- After change value :
image

# Result for log level : 
image

# Result for Background image: 
image
if i set image path no exist get so now, more correct compared before
image

# Game Background exemple:
image 
image

# Résultat for see game background switch : 
https://youtu.be/VOVSrVXe6eU

### CODE REVIEW

**Comment by Luffykun007** [comment]
@Zangetsu the background image does not work for me, what am I doing wrong?

Capturar12

**Comment by 1whatleytay** [comment]
@Luffykun007 Do you actually have an image called "2.jpg" in the same folder as Vita3K? You might have to put a full path to the image you want to use like `C:\Users\Luffy\Desktop\smiley.png`.

**Comment by Luffykun007** [comment]
@1whatleytay Okay, I'll try this.

**Comment by Zangetsu38** [comment]
Yes, i see your image but if you d'ont have édit emulatot path for indicate root vita3k fonder is normal so here it look on user/roaming/vita3k/vita3k folder look my first screen for préf path

**Comment by Luffykun007** [comment]
@Zangetsu the path of the emulator seems this correct, but I continue with the same error here.
also did not work previous suggestion of @ 1whatleytay.

Capturar13
Capturar14

**Comment by 1whatleytay** [comment]
Can you do my suggestion and show me the console? It'll print out a special error message on why it's not showing the image (there are two possible messages :P).

Make sure you provide a good image and that the path doesn't have some weird extra letters or anything. Double check the path by right clicking the image and going to properties.

**Comment by Zangetsu38** [comment]
i see so here is not correect, show your path with Vita3K .exe is normally totaly diferent rout pref path

**Comment by Luffykun007** [comment]
@ 1whatleytay I do not know if that's it, but it's there.
PS: In the way of the emulator not there is this Luffy but my real username, Luffy is just to illustrate.

Image1

Capturar16

**Comment by Zangetsu38** [comment]
@Luffykun007 your pathé folder is wrong here, deskop is not path for your Vita3K exe, look your picture, here is wrong and show your path for Vita3K using

**Comment by 1whatleytay** [comment]
@Zangetsu38 is right. The path I gave you was kinda an example, it's to an image called "smiley.png" on your desktop, but in your screenshot it appears to be in your Vita3K exe folder. I bet if you drag 3.png to your desktop it will work.

**Comment by Zangetsu38** [comment]
No, path for image used no root ".exe", but pref path folder with so ux0 folder, so if you want just write name image put image inside your pref path folder, likeme on my screen for Pref path in D:/emulator/Vita3K

**Comment by Luffykun007** [comment]
@ 1whatleytay I dragged the image to the desktop but it also did not work .... sorry guys, I will keep trying here and if I can not then I will open a new problem for this.

PS:PS:@Zangetsu could you send me your config.yml with your configuration,could be some something wrong here.

**Comment by Zangetsu38** [comment]
go on discord for talk

**Comment by Luffykun007** [comment]
Okay

**Comment by Zangetsu38** [comment]
This pr is ready for review

**Comment by Zangetsu38** [comment]
pushed with fix crash for debug build if try return default background without user background set.

**Comment by Zangetsu38** [comment]
@Luffykun007 rested last build please
fix crash caused without user background and with change user background.
Code little reworked for init_background for get all works good

File: src/emulator/gui/include/gui/state.h [hunk 1]
```diff
@@ -93,7 +93,9 @@ struct State {
     char disassembly_count[5] = "100";
     std::vector<std::string> disassembly;
 
-    GLObject background_texture;
+    GLuint current_background = 0;
+    std::map<std::string, GLObject> game_backgrounds;
+    std::map<std::string, GLObject> user_backgrounds;
```
Comment (1whatleytay) [inline src/emulator/gui/include/gui/state.h]: You probably don't need a std::map for `user_backgrounds`. 😓 When you set a new background, there isn't any reason to keep the old background in memory (you probably aren't going to use it again).

But I see why you did it, you can't change GLObject once it was created (init freaks out). I would add a method to GLObject so we can replace the texture (deletes old, sets new), but if someone else has a better idea, please tell.

File: src/emulator/gui/src/gui.cpp [hunk 2]
```diff
@@ -142,9 +142,9 @@ static void init_font(State &gui) {
     gui.normal_font = io.Fonts->AddFontFromMemoryTTF(gui.font_data.data(), static_cast<int>(font_file_size), 16, &font_config, io.Fonts->GetGlyphRangesJapanese());
 }
 
-static void init_background(State &gui, const std::string &image_path) {
+void init_background(HostState &host, const std::string &image_path) {
```
Comment (1whatleytay) [inline src/emulator/gui/src/gui.cpp]: Does `init_background` use `HostState` or just `gui::State`? Not too sure why this change was made.

File: src/emulator/gui/src/settings_dialog.cpp [hunk 3]
```diff
@@ -75,12 +86,29 @@ void draw_settings_dialog(HostState &host) {
...
+            } else if (host.gui.user_backgrounds[host.cfg.background_image]) {
+                host.gui.current_background = host.gui.user_backgrounds[host.cfg.background_image];
+            }
+        }
+        ImGui::Spacing();
+        if (host.gui.current_background) {
+            ImGui::SliderFloat("Background Alpha \nSelect your preferred transparent background effect.", &host.cfg.background_alpha, 0.999f, 0.000f);
+            if (ImGui::IsItemHovered())
+                ImGui::SetTooltip("Minimum slider is opaque and maximum is transparent.");
+        }
```
Comment (1whatleytay) [inline src/emulator/gui/src/settings_dialog.cpp]: There are some weird grammar issues with some of the tool tips. Don't be afraid to use a tool like Grammarly or the Internet! I do it all the time.

File: src/emulator/config.cpp [hunk 4]
```diff
@@ -96,17 +96,18 @@ bool serialize(Config &cfg) {
...
+    config_file_emit_single(emitter, "game-background", cfg.game_background);
     config_file_emit_single(emitter, "icon-size", cfg.icon_size);
     config_file_emit_single(emitter, "archive-log", cfg.archive_log);
     config_file_emit_single(emitter, "texture-cache", cfg.texture_cache);
     config_file_emit_single(emitter, "sys-button", cfg.sys_button);
     config_file_emit_single(emitter, "sys-lang", cfg.sys_lang);
+    config_file_emit_single(emitter, "background-image", cfg.background_image);
     config_file_emit_single(emitter, "background-alpha", cfg.background_alpha);
+    config_file_emit_single(emitter, "log-level", cfg.log_level);
+    config_file_emit_single(emitter, "pref-path", cfg.pref_path);
```
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply src/emulator/config.cpp]: For can set all setting inside setting dialog is obligatory because it have function inside config

File: src/emulator/gui/src/gui.cpp (hunk 2 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/gui.cpp]: For get works on settings dialog ^^ and use like this same other works perfectly

File: src/emulator/gui/include/gui/state.h (hunk 1 above)
Comment (Zangetsu38) [reply src/emulator/gui/include/gui/state.h]: For me here is juste better  and is need obligatoty separate game and user background for all works

File: src/emulator/config.cpp [hunk 5]
```diff
@@ -96,17 +96,18 @@ bool serialize(Config &cfg) {
     config_file_emit_single(emitter, "log-uniforms", cfg.log_uniforms);
     config_file_emit_single(emitter, "pstv-mode", cfg.pstv_mode);
     config_file_emit_single(emitter, "show-gui", cfg.show_gui);
+    config_file_emit_single(emitter, "game-background", cfg.game_background);
```
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply src/emulator/config.cpp]: Yes is that i have said you i have remove code by accident lol i restore later

File: src/emulator/gui/src/settings_dialog.cpp (hunk 3 above)
Comment (Luffykun007) [reply src/emulator/gui/src/settings_dialog.cpp]: @Zangetsu Good job, man! Background of the game is working now, I still can not load a custom image, there must be something wrong here in my setup ... but this is already very good, thank you guys.

Capturar2
Capturar1

**Comment by Luffykun007** [comment]
@Zangetsu Good job, man! Background of the game is working now, I still can not load a custom image, there must be something wrong here in my setup ... but this is already very good, thank you guys.
Capturar2
Capturar1

File: src/emulator/config.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply src/emulator/config.cpp]: reset it :)

File: src/emulator/config.cpp [hunk 6]
```diff
@@ -135,16 +136,17 @@ static bool deserialize(Config &cfg) {
...
     get_yaml_value(config_node, "archive-log", &cfg.archive_log, false);
     get_yaml_value(config_node, "texture-cache", &cfg.texture_cache, true);
     get_yaml_value(config_node, "sys-button", &cfg.sys_button, static_cast<int>(SCE_SYSTEM_PARAM_ENTER_BUTTON_CROSS));
     get_yaml_value(config_node, "sys-lang", &cfg.sys_lang, static_cast<int>(SCE_SYSTEM_PARAM_LANG_ENGLISH_US));
+    get_yaml_value(config_node, "background-image", &cfg.background_image, static_cast<std::string>(""));
     get_yaml_value(config_node, "background-alpha", &cfg.background_alpha, 0.000f);
-    get_yaml_value_optional(config_node, "log-level", &cfg.log_level, static_cast<int>(spdlog::level::trace));
-    get_yaml_value_optional(config_node, "pref-path", &cfg.pref_path);
+    get_yaml_value(config_node, "log-level", &cfg.log_level, static_cast<int>(spdlog::level::trace));
+    get_yaml_value(config_node, "pref-path", &cfg.pref_path, static_cast<std::string>(""));
```
Comment (pent0) [inline src/emulator/config.cpp]: The fact that this work is amazing

```suggestion
    get_yaml_value(config_node, "pref-path", &cfg.pref_path, std::string{});
```

File: src/emulator/gui/src/gui.cpp [hunk 7]
```diff
@@ -142,9 +142,9 @@ static void init_font(State &gui) {
...
     if (!fs::exists(image_path)) {
-        LOG_INFO("Invalid background file path {}.", image_path);
+        LOG_WARN("Image not found from {}, check your image path set.", image_path);
```
Comment (pent0) [inline src/emulator/gui/src/gui.cpp]:
```suggestion
        LOG_WARN("Image doesn't exist {}.", image_path);
```

Simplify the statement

File: src/emulator/gui/src/gui.cpp [hunk 8]
```diff
@@ -153,11 +153,12 @@ static void init_background(State &gui, const std::string &image_path) {
     stbi_uc *data = stbi_load(image_path.c_str(), &width, &height, nullptr, STBI_rgb_alpha);
 
     if (!data) {
-        LOG_INFO("Could not load image from {}.", image_path);
+        LOG_ERROR("Invalide format image from {}, check your image format.", image_path);
```
Comment (pent0) [inline src/emulator/gui/src/gui.cpp]: What's with this weird French append

```suggestion
        LOG_ERROR("Invalid image format or corrupted image: {}.", image_path);
```

File: src/emulator/gui/src/settings_dialog.cpp (hunk 3 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: i have fixed normally

**Comment by Zangetsu38** [comment]
Push finnaly fix for return default background after get cursor on game, and add also news Button for can reset background user setting and return on default.
@Luffykun007 test if you can switch game background/default background 
return on default with set cursor on colunm Version/Tittle.

File: src/emulator/gui/src/gui.cpp [hunk 9]
```diff
@@ -153,12 +153,12 @@ static void init_background(State &gui, const std::string &image_path) {
     stbi_uc *data = stbi_load(image_path.c_str(), &width, &height, nullptr, STBI_rgb_alpha);
 
     if (!data) {
-        LOG_INFO("Could not load image from {}.", image_path);
+        LOG_ERROR("Image is Invalide format or corrupted: {}.", image_path);
```
Comment (pent0) [inline src/emulator/gui/src/gui.cpp]:
```suggestion
        LOG_ERROR("Image has invalid format or corrupted: {}.", image_path);
````

File: src/emulator/gui/src/settings_dialog.cpp [hunk 10]
```diff
@@ -56,13 +59,21 @@ void draw_settings_dialog(HostState &host) {
...
         ImGui::Checkbox("Archive Log", &host.cfg.archive_log);
         ImGui::SameLine();
         if (ImGui::IsItemHovered())
             ImGui::SetTooltip("Check the box to enable Archiving Log.");
         ImGui::Checkbox("Texture Cache", &host.cfg.texture_cache);
         if (ImGui::IsItemHovered())
             ImGui::SetTooltip("Uncheck the box to disable texture cache.");
+        ImGui::Spacing();
+        ImGui::PushItemWidth(400);
+        ImGui::InputTextWithHint("Set your Path Emulator \n(Reboot after change for apply)", "Write your path folder here", &host.cfg.pref_path);
```
Comment (pent0) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
        ImGui::InputTextWithHint("Set emulated system storage folder \n(Reboot after change for apply)", "Write the path to the folder here", &host.cfg.pref_path);
```

This is the best i can think of

File: src/emulator/gui/src/settings_dialog.cpp [hunk 11]
```diff
@@ -56,13 +59,21 @@ void draw_settings_dialog(HostState &host) {
...
+        ImGui::PopItemWidth();
+        if (ImGui::IsItemHovered())
+            ImGui::SetTooltip("For can use personal folder for Vita3K. \nCopy and past here your path Vita3K folder.");
```
Comment (pent0) [inline src/emulator/gui/src/settings_dialog.cpp]: At least try to check your grammar, this is just unacceptable, we are writing software for general use.

```suggestion
            ImGui::SetTooltip("Store game data in your personal folder. \nAdd the path to the folder here.");
```

File: src/emulator/gui/src/settings_dialog.cpp [hunk 12]
```diff
@@ -75,12 +86,34 @@ void draw_settings_dialog(HostState &host) {
...
+        if (ImGui::IsItemHovered())
+            ImGui::SetTooltip("Uncheck the box to disable viewing Game background.");
         ImGui::Spacing();
         ImGui::SliderInt("Game Icon Size \nSelect your preferred icon size.", &host.cfg.icon_size, 16, 128);
         ImGui::Spacing();
-        ImGui::SliderFloat("Background Alpha \nSelect your preferred transparent background effect.", &host.cfg.background_alpha, 0.999f, 0.000f);
-        if (ImGui::IsItemHovered())
-            ImGui::SetTooltip("Minimum slider is opaque and maximum is transparent.");
+        ImGui::PushItemWidth(400);
+        ImGui::InputTextWithHint("Set your Path for Background Image", "Write your path image here", &host.cfg.background_image);
```
Comment (pent0) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
        ImGui::InputTextWithHint("Set background image", "Add your path to the image here", &host.cfg.background_image);
```

File: src/emulator/gui/src/gui.cpp (hunk 9 above)
Comment (1whatleytay) [reply src/emulator/gui/src/gui.cpp]: Or even better: "Invalid or corrupted image: {}"

Short and sweet.

**Comment by Zangetsu38** [comment]
Change typo pushed

**Comment by Luffykun007** [comment]
Yeah, works for me @Zangetsu.
